### PR TITLE
Fix/Add First Person Attachments

### LINKED ISSUE #10010: Entities disappear in 3rd person mode when running set_attach()
##### Minetest version
```
Minetest 5.3.0-dev-8fc9e7eb1 (Windows)
```

##### OS / Hardware
Operating system: Windows 10
CPU: Ryzen 3200G

GPU model: 8192MB ATI Radeon RX 580 Series (XFX Pine Group)
OpenGL version: 4.6

##### Summary
Entities do not appear/disappear in 3rd person mode when running set_attach() to a player.

##### Steps to reproduce
Run code which runs ```entity:set_attach(player,"Body",vector.new(0,0,0),vector.new(0,0,0)) ``` and they will disappear until camera change.

##### Fun Fact
10010 is binary for ```?```

### PR DESCRIPTION
**Pull request by Jordach**
Adds First Person Attachments, 

- Adopts and fixes #8342 
- Fixes #10010 
- Fixes #8046
- Fixes some other third person camera specific attachments.
- Implements a single new flag for entities to be forced visible in first person mode.
- Old mods do not need to be updated to use the new flag and are fully backwards compatible.

## To do
~~Fix players appearing in first person.~~

Ready for Review.

## How to test

```lua
set_attach(object_ref, "your_bone_here", vector.new(), vector.new(), true) -- Enable first person attachments for this entity.
set_attach(object_ref, "your_bone_here", vector.new(), vector.new(), false) -- Disable first person attachments for this entity.
set_attach(object_ref, "your_bone_here", vector.new(), vector.new()) -- Disable first person attachments for this entity.
```

Everything such as `set_animation` and `set_bone_position` work even in first person mode.

## Screenshots / Media

https://gfycat.com/timelybrokenbighornedsheep
https://gfycat.com/welloffillfatedemperorshrimp

### CODE REVIEW

**Comment by Jordach** [comment]
Fixed the players being shown, as entities now show their position in world proper:
https://gfycat.com/hiddenimaginarybushsqueaker

File: src/client/content_cao.cpp [hunk 1]
```diff
@@ -1865,5 +1884,33 @@ std::string GenericCAO::debugInfoText()
 	return os.str();
 }
 
+void GenericCAO::hideMeshTexture()
```
Comment (SmallJoker) [inline src/client/content_cao.cpp:1887]: Does not only hide, but also recover visibility. Suggested naming:
`void GenericCAO::updateTextureVisibility()`

File: src/script/lua_api/l_object.cpp [hunk 2]
```diff
@@ -702,7 +703,10 @@ int ObjectRef::l_set_attach(lua_State *L)
 	rotation = v3f(0, 0, 0);
 	if (!lua_isnil(L, 5))
 		rotation = read_v3f(L, 5);
-	co->setAttachment(parent->getId(), bone, position, rotation);
+	force_visible = false;
+	if (!lua_isnil(L, 6))
+		force_visible = readParam<bool>(L, 6);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:708]:
```suggestion
	force_visible = readParam<bool>(L, 6, false);
```

File: src/server/unit_sao.cpp [hunk 3]
```diff
@@ -121,8 +121,8 @@ void UnitSAO::sendOutdatedData()
 }
 // clang-format on
 
-void UnitSAO::setAttachment(
-		int parent_id, const std::string &bone, v3f position, v3f rotation)
+void UnitSAO::setAttachment(int parent_id, const std::string &bone,
+			v3f position, v3f rotation, bool force_visible)
```
Comment (SmallJoker) [inline src/server/unit_sao.cpp:125]: Two indents only, please.

File: src/server/unit_sao.cpp [hunk 4]
```diff
@@ -145,21 +146,22 @@ void UnitSAO::setAttachment(
 	}
 }
 
-void UnitSAO::getAttachment(
-		int *parent_id, std::string *bone, v3f *position, v3f *rotation) const
+void UnitSAO::getAttachment(int *parent_id, std::string *bone, v3f *position,
+			v3f *rotation, bool *force_visible) const
```
Comment (SmallJoker) [inline src/server/unit_sao.cpp:150]: Two indents only, please.

File: doc/lua_api.txt [hunk 5]
```diff
@@ -6097,11 +6097,13 @@ object you are working with still exists.
...
     * `frame_speed`: number, default: `15.0`
-* `set_attach(parent, bone, position, rotation)`
+* `set_attach(parent, bone, position, rotation, forced_visible)`
     * `bone`: string
     * `position`: `{x=num, y=num, z=num}` (relative)
     * `rotation`: `{x=num, y=num, z=num}` = Rotation on each axis, in degrees
-* `get_attach()`: returns parent, bone, position, rotation or nil if it isn't
+    * `forced_visible`: Boolean to control whether the attached entity 
+       should appear in first person.
+* `get_attach()`: returns parent, bone, position, rotation, forced_visible, or nil if it isn't
```
Comment (SmallJoker) [inline doc/lua_api.txt:6106]: Line needs breaking (>80 characters)

File: src/script/lua_api/l_object.cpp [hunk 6]
```diff
@@ -731,6 +736,7 @@ int ObjectRef::l_get_attach(lua_State *L)
 	lua_pushlstring(L, bone.c_str(), bone.size());
 	push_v3f(L, position);
 	push_v3f(L, rotation);
+	lua_pushboolean(L, force_visible);
 	return 4;
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:740]:
```suggestion
	return 5;
```

**Review by SmallJoker — COMMENTED** [review]
Character looks like this after taking damage (or applying any other texture changes):

grafik

File: src/server/luaentity_sao.cpp [hunk 7]
```diff
@@ -240,7 +240,7 @@ std::string LuaEntitySAO::getClientInitializationData(u16 protocol_version)
 		msg_os << serializeLongString(generateUpdateBonePositionCommand(
 			bone_pos.first, bone_pos.second.X, bone_pos.second.Y)); // m_bone_position.size
 	}
-	msg_os << serializeLongString(generateUpdateAttachmentCommand()); // 4
+	msg_os << serializeLongString(generateUpdateAttachmentCommand()); // 5
```
Comment (SmallJoker) [inline src/server/luaentity_sao.cpp:243]: Erroneous change. If even - this should be changed to `4 + m_bone_position.size`

File: src/server/player_sao.cpp [hunk 8]
```diff
@@ -124,7 +124,7 @@ std::string PlayerSAO::getClientInitializationData(u16 protocol_version)
 		msg_os << serializeLongString(generateUpdateBonePositionCommand(
 			bone_pos.first, bone_pos.second.X, bone_pos.second.Y)); // m_bone_position.size
 	}
-	msg_os << serializeLongString(generateUpdateAttachmentCommand()); // 4
+	msg_os << serializeLongString(generateUpdateAttachmentCommand()); // 5
```
Comment (SmallJoker) [inline src/server/player_sao.cpp:127]: Incorrect change. If even - this should be changed to `4 + m_bone_position.size`

**Review by SmallJoker — APPROVED** [review]
PR works. All entities appear correctly, even when the player takes damage.

**Review by appgurueu — APPROVED** [review]

File: src/client/content_cao.cpp [hunk 9]
```diff
@@ -456,7 +456,11 @@ void GenericCAO::setChildrenVisible(bool toset)
 	for (u16 cao_id : m_attachment_child_ids) {
 		GenericCAO *obj = m_env->getGenericCAO(cao_id);
 		if (obj) {
-			obj->setVisible(toset);
+			// Check if the entity is forced to appear in first person.
+			if (obj->isForcedVisible())
+				obj->setVisible(obj->isForcedVisible());
+			else
+				obj->setVisible(toset);
```
Comment (sfan5) [inline src/client/content_cao.cpp:463]:
```suggestion
			obj->setVisible(obj->isForcedVisible() ? true : toset);
```

File: src/client/content_cao.cpp [hunk 10]
```diff
@@ -1865,5 +1884,33 @@ std::string GenericCAO::debugInfoText()
...
+
+	// Grab the active player scene node so we know there's
+	// at least a mesh to occlude from the camera.
+	irr::scene::ISceneNode *node = getSceneNode();
+	if (!node)
+		return;
+
+	if (m_client->getCamera()->getCameraMode() == CAMERA_MODE_FIRST) {
+		// Hide the mesh by culling faces with normals facing
+		// the camera. Serious hackyness but it works for our
```
Comment (sfan5) [inline src/client/content_cao.cpp:1900]:
```suggestion
		// Hide the mesh by culling both front and
		// back faces. Serious hackyness but it works for our
```

**Review by sfan5 — COMMENTED** [review]
Looks good other than that, but haven't tested.

File: src/client/content_cao.cpp (hunk 10 above)
Comment (appgurueu) [reply src/client/content_cao.cpp:1900]: Consider annotating this as `// HACK Hide the mesh by culling faces with normals facing the camera`

**Comment by sfan5** [comment]
Since commit 65c15e137fe584edb38edea21c49873be00d554c was merged this PR also needs up be updated:

https://github.com/minetest/minetest/pull/10360/checks?check_run_id=1186523409#step:4:522

**Review by sfan5 — APPROVED** [review]
Tested, works.

**Pull request merged**